Distribution/Server/Features/ReverseDependencies/State.hs

```diff
@@ -13,406 +15,203 @@ import Distribution.PackageDescription
-    goUnregister prev (pkgname, (range, versions)) =
-        let revPackage = Map.fromList $ map (\v -> (v, pkgMap)) versions
-            revRange   = Map.singleton pkgname (Map.fromList $ map (\v -> (v, range)) versions)
-        -- there are possibly better ways to go about this
-        in  Map.differenceWith (\(a, b) (c, d) -> keepMaps $
-              ( Map.differenceWith (\e f ->
-                    keepMap $ Map.differenceWith (\g h ->
-                        keepSet $ Set.difference g h)
-                    e f)
-                a c
-              , Map.differenceWith (\e f ->
-                    keepMap $ Map.difference e f)
-                b d
-              )) prev (Map.singleton pkgname (revPackage, revRange))
-    backtrace prev = Map.update (\(vs, rs) -> keepMaps (Map.delete version vs, rs)) name prev
-
---------------------------------------------------------------------------------
--- Calculating dependencies and selecting versions
-
--- | Given a package, determine the packages on which it depends.
--- For all such packages, return the specific versions that satisfy the
--- dependency as indicated in the cabal file.
-getLinkedNodes :: (PackageName -> [Version])
-               -> CombinedDeps -> Map PackageName [Version]
-getLinkedNodes getVersions pkgs = Map.mapWithKey (\pkg range -> selectVersions range $ getVersions pkg) pkgs
-
--- | Given a dependency (a package name and a version range), find all versions
--- in the current package index that satisfy it.
-selectVersions :: VersionRange -> [Version] -> [Version]
-selectVersions range versions= filter (flip withinRange range) versions
-
--- | Collect all dependencies specified in a package's cabal file, considering
--- all alternatives.
---
--- This unions all version ranges together from different branches, which is
--- imprecise but mostly good enough (a slightly better heuristic might be
--- intersecting within a block, but unioning blocks together).
-getAllDependencies :: PkgInfo -> CombinedDeps
-getAllDependencies pkg =
-    let desc = pkgDesc pkg
-    in Map.fromListWith unionVersionRanges $ toDepsList (maybeToList $ condLibrary desc)
-                                          ++ toDepsList (map snd $ condExecutables desc)
-  where toDepsList :: [CondTree v [Dependency] a] -> [(PackageName, VersionRange)]
-        toDepsList l = [ (p, v) | Dependency p v <- concatMap harvestDependencies l ]
-
--- | Collect all dependencies from all branches of a condition tree.
-harvestDependencies :: CondTree v [Dependency] a -> [Dependency]
-harvestDependencies (CondNode _ deps comps) = deps ++ concatMap forComponent comps
-  where forComponent (_, iftree, elsetree) = harvestDependencies iftree ++ maybe [] harvestDependencies elsetree
+constructRevDeps :: PackageIndex PkgInfo -> Bimap PackageId NodeId -> RevDeps
+constructRevDeps index nodemap =
+    let allPackages = concatMap (take 5) $ PackageIndex.allPackagesByName index
```

Comment:
Any idea what this take 5 is about? Is it trying to take the 5 latest versions? Are we sure it's not actually getting the 5 oldest?